File: Python_TRR/1.py
# Viết bằng Python
# Bài 1: VSinh tổ hợp (Combination).

# Giải thích: Tổ hợp (Combination) là cách chọn ra r phần tử từ n phần tử mà không quan tâm đến thứ tự.

from itertools import combinations # Sử dụng hàm combinations từ thư viện itertools.

def generate_combinations(arr, r): # Hàm sinh tổ hợp. 
    return list(combinations(arr, r)) # Trả về danh sách các tổ hợp.

# Ví dụ sử dụng
arr = [1, 2, 3, 4] # Tập hợp ban đầu.
r = 2 # Số phần tử của tổ hợp.
print("Tổ hợp:")
for comb in generate_combinations(arr, r):
    print(comb)

# Kết quả
# Tổ hợp:
# (1, 2)
# (1, 3)
# (1, 4)
# (2, 3)
# (2, 4)
# (3, 4)

File: C_TRR/1.c
// Viết bằng C
// Bài 1: Sinh tổ hợp (Combination).

#include <stdio.h>

void printCombination(int arr[], int n, int r, int index, int data[], int i) 
{
    if (index == r) 
    {
        for (int j = 0; j < r; j++)
            printf("%d ", data[j]);
        printf("\n");
        return;
    }

    if (i >= n)
        return;

    data[index] = arr[i];
    printCombination(arr, n, r, index + 1, data, i + 1);
    printCombination(arr, n, r, index, data, i + 1);
}

void generateCombinations(int arr[], int n, int r) 
{
    int data[r];
    printCombination(arr, n, r, 0, data, 0);
}

int main() 
{
    int arr[] = {1, 2, 3, 4};
    int n = 4, r = 2;
    printf("Tổ hợp:\n");
    generateCombinations(arr, n, r);
    return 0;
}

